File: python/leetcode/250.py
class Solution:
    def countUnivalSubtrees(self, root):
        self.count = 0
        self.unival(root)
        return self.count

    def unival(self, root):
        if not root: return True
        l = self.unival(root.left)
        r = self.unival(root.right)
        if l and r and (not root.left or root.left.val == root.val) and (not root.right or root.right.val == root.val):
            self.count += 1
            return True
        return False

File: cpp/leetcode/250.cpp
class Solution {
public:
    int countUnivalSubtrees(TreeNode* root) {
        int count = 0;
        help(root, count);
        return count;
    }

    bool help(TreeNode* root, int& count) {
        if(!root) return true;
        auto isLeftUnival = help(root->left, count);
        auto isRightUnival = help(root->right, count);
        if(isLeftUnival && isRightUnival
                && ((!root->left) || root->left->val == root->val)
                && ((!root->right) || root->right->val == root->val)
          ) {
            ++count;
            return true;
        }
        return false;
    }
};